man/man1/xpaaccess.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "xpaaccess 1"
.TH xpaaccess 1 "July 23, 2013" "version 2.1.15" "SAORD Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
\&\fBxpaaccess: see if template matches registered \s-1XPA\s0 access points\fR
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
xpaaccess [\-c] [\-h] [\-i nsinet] [\-m method] [\-n] [\-t sval,lval] [\-u users] \-v <template> [type]
.SH "OPTIONS"
.IX Header "OPTIONS"
.Vb 10
\&  \-c            contact each access point individually
\&  \-h            print help message
\&  \-i            access XPA point on different machine (override XPA_NSINET)
\&  \-m            override XPA_METHOD environment variable
\&  \-n            return number of matches instead of "yes" or "no"
\&  \-t [s,l]      set short and long timeouts (override XPA_[SHORT,LONG]_TIMEOUT)
\&  \-u [users]    XPA points can be from specified users (override XPA_NSUSERS)
\&  \-v            print info about each successful access point
\&  \-V            print info or error about each access point
\&  \-\-version     display version and exit
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
xpaaccess returns \*(L"yes\*(R" to stdout (with a return error code if 1) if there are
existing \s-1XPA\s0 access points that match the 
template
(and optional access type: g,i,s). Otherwise, it returns \*(L"no\*(R" (with a
return error code of 0).  If \-n is specified, the number of matches is
returned instead (both to stdout and in the returned error code). If
\&\-v is specified, each access point is displayed to stdout instead of
the number of matches.
.PP
By default, xpaaccess simply contacts the xpans name server to find
the list of registered access points that match the specified
template. It also checks to make sure the specified types are
supported by that access point. This is the fastest way to determine
available access points. However, an access point might registered but
not yet available, if, for example, the server program has not entered
its event loop to process \s-1XPA\s0 requests. To find access points that are
guaranteed to be available for processing, use the \-c (contact)
switch.  With this switch, xpaaccess contacts each matching \s-1XPA\s0 server
(rather than the name server) to make sure the registered access point
really is ready for processing. In this mode, if an access point is
registered but not available, xpaaccess will pause for a period of
time equal to the \s-1XPA_LONG_TIMEOUT\s0, in order to give the server a
chance to ready itself. By default, this timeout is 30 seconds. You
can shorten the time of delay using the \-t \*(L"short,long\*(R" switch. For
example, to shorten the delay time to 2 seconds, use:
.PP
.Vb 1
\&  xpaaccess \-c \-t "2,2" ds9
.Ve
.PP
The first argument is the short delay value, and is ignored in this
operation. The second is the long delay timeout.
.PP
Note also that the default xpaaccess method (no \-c switch) does not
check access control (acls) but rather only checks whether the access
point is both registered with the xpans name server and provides the
specified type of access. In other words, the default xpaaccess could
return 'yes' when you might not actually have access. This mode also
always returns 'yes' for the xpans name server itself, regardless of
whether the name server is active. The \-c (contact) switch, which
contacts the access point directly, can and does check the access
control (only for servers using version 2.1 and above) and also
returns the real status of xpans.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
See xpa(n) for a list of \s-1XPA\s0 help pages


xpaaccess.c
/*
 *	Copyright (c) 1999-2003 Smithsonian Astrophysical Observatory
 */

#include <xpap.h>

#define MAX_XPAS 10000

extern char *optarg;
extern int optind;

#ifdef ANSI_FUNC
void
usage (char *s)
#else
void usage(s)
     char *s;
#endif
{
    fprintf(stderr, "\n");
    fprintf(stderr, "usage:\n");
    fprintf(stderr, "  %s [-c] [-h] [-i nsinet] [-m method] [-n] [-t sval,lval] [-u users] [-v|-V] <template> [type]\n", s);
    fprintf(stderr, "\n");
    fprintf(stderr, "switches:\n");
    fprintf(stderr, "\t-c\tcontact each access point individually to see if it is available\n");
    fprintf(stderr, "\t-h\tprint this message\n");
    fprintf(stderr, "\t-i\toverride XPA_NSINET environment variable\n");
    fprintf(stderr, "\t-m\toverride XPA_METHOD environment variable\n");
    fprintf(stderr, "\t-n\treturn number of matches instead of \"yes\" or \"no\"\n");
    fprintf(stderr, "\t-t \toverride XPA_[SHORT,LONG]_TIMEOUT environment variables\n");
    fprintf(stderr, "\t-u \toverride XPA_NSUSERS environment variable\n");
    fprintf(stderr, "\t-v\tprint info about successful access points\n");
    fprintf(stderr, "\t-V\tprint info and errors about all access points\n");
    fprintf(stderr, "\t--version display version and exit\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "xpaaccess returns \"yes\" if there are existing XPA access points that match\n");
    fprintf(stderr, "the template (and optional access type: g,i,s), and returns \"no\" otherwise.\n");
    fprintf(stderr, "If -n is specified, the number of matches is returned.\n");
    fprintf(stderr, "Templates are of the form [class:]name. Wildcards *,?,[] are accepted.\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "examples:\n");
    fprintf(stderr, "\tcsh> if( `xpaaccess ds9` == \"yes\") then ...\n" );
    fprintf(stderr, "\tcsh> set got=`xpaaccess -n ds9 g`\n");
    fprintf(stderr, "\n(version: %s)\n", XPA_VERSION);
    exit(0);
}

/* catch error from pre 2.1 server -- if we find this error, we know the
   access point is available */
#define OLD_SERVER(s) strstr(s, "XPA$ERROR invalid xpa command in initialization string")

#ifdef ANSI_FUNC
int 
main (int argc, char **argv)
#else
int
main(argc, argv)
     int argc;
     char **argv;
#endif
{
  int i;
  int c;
  int got;
  int n=0;
  int donum=0;
  int verbosity=0;
  int contact=0;
  char cmd[SZ_LINE];
  char mode[SZ_LINE];
  char *type=NULL;
  char *xtmpl=NULL;
  char **rclasses;
  char **rnames;
  char **rmethods;
  char **rinfos;
  char *names[MAX_XPAS];
  char *errs[MAX_XPAS];
  XPA xpa=NULL;

  /* display version and exit, if necessary */
  if( (argc == 2) && !strcmp(argv[1], "--version") ){
    fprintf(stderr, "%s\n", XPA_VERSION);
    exit(0);
  }

  /* start with no mode flag */
  *mode = '\0';

  /* we want the args in the same order in which they arrived, and
     gnu getopt sometimes changes things without this */
  putenv("POSIXLY_CORRECT=true");

  /* process switch arguments */
  while ((c = getopt(argc, argv, "chi:m:nt:u:vVwW")) != -1){
    switch(c){
    case 'c':
      contact=1;
      break;
    case 'h':
      usage(argv[0]);
    case 'i':
      snprintf(cmd, SZ_LINE, "XPA_NSINET=%s", optarg);
      putenv(xstrdup(cmd));
      break;
    case 'm':
      snprintf(cmd, SZ_LINE, "XPA_METHOD=%s", optarg);
      putenv(xstrdup(cmd));
      break;
    case 'n':
      donum=1;
      break;
    case 't':
      {
	int xip=0;
	char xbuf[SZ_LINE];
	newdtable(",;");
	if( word(optarg, xbuf, &xip) && *xbuf && (*xbuf != '*') ){
	  snprintf(cmd, SZ_LINE, "XPA_SHORT_TIMEOUT=%s", xbuf);
	  putenv(xstrdup(cmd));
	}
	if( word(optarg, xbuf, &xip) && *xbuf && (*xbuf != '*') ){
	  snprintf(cmd, SZ_LINE, "XPA_LONG_TIMEOUT=%s", xbuf);
	  putenv(xstrdup(cmd));
	}
	freedtable();
      }
      break;
    case 'u':
      snprintf(cmd, SZ_LINE, "XPA_NSUSERS=%s", optarg);
      putenv(xstrdup(cmd));
      break;
    case 'v':
      verbosity=1;
      break;
    case 'V':
      verbosity=2;
      break;
    /* for backward compatibility with 1.0 */
    case 'w':
    case 'W':
      break;
    }
  }

  /* we must have a template */
  if( optind >= argc ){
    usage(argv[0]);
  }
  else{
    xtmpl = argv[optind];
    optind++;
    if( optind < argc ){
      type = argv[optind];
    }
  }

  /* if we need to contact each server ... */
  if( contact ){
    got = XPAAccess(xpa, xtmpl, type, mode, names, errs, MAX_XPAS);
    for(i=0; i<got; i++){
      if( !errs[i] || OLD_SERVER(errs[i]) ){
	if( verbosity ) fprintf(stdout, "%s\n", names[i]);
	n++;
      }
      else{
	if( verbosity >= 2 ) fprintf(stdout, "%s", errs[i]);
      }
      if( errs[i] )  xfree(errs[i]);
      if( names[i] ) xfree(names[i]);
    }
  }
  /* only contact name server */
  else{
    n = XPANSLookup(xpa, xtmpl, type, &rclasses, &rnames, &rmethods, &rinfos);
    /* free up the space */
    for(i=0; i<n; i++){
      if( verbosity ){
	fprintf(stdout, "%s", rmethods[i]);
	if( strcmp(rinfos[i], XPA_DEF_CLIENT_INFO) )
	  fprintf(stdout, " %s", rinfos[i]);
	fprintf(stdout, "\n");
      }
      /* done with these strings */
      xfree(rclasses[i]);
      xfree(rnames[i]);
      xfree(rmethods[i]);
      xfree(rinfos[i]);
    }
    /* free up arrays returned by name server */
    if( n > 0 ){
      xfree(rclasses);
      xfree(rnames);
      xfree(rmethods);
      xfree(rinfos);
    }
  }

  /* print out number */
  if( !verbosity && donum ){
    fprintf(stdout, "%d\n", n);
  }
  /* print out yes/no */
  else if( !verbosity && !donum ){
    if( n > 0 )
      fprintf(stdout, "yes\n");
    else
      fprintf(stdout, "no\n");
  }
  fflush(stdout);
  return(n);
}
